Setup browse endpoint for browsing Directus extensions

rijkvanzanten opened the pull request:
## Scope

What's changed:

- Adds a new `@directus/extensions-registry` package with functions to explore the registry for Directus extensions
- Adds a new `/extensions/registry` endpoint to the API to browse available marketplace extensions
- Adds a new `/extensions/registry/:id` endpoint to get more detailed information about a single available marketplace extension
- Adds a new OutOfDateError to the errors abstraction to throw when using the marketplace registry with an outdated copy of Directus 

## Potential Risks / Drawbacks

- New logic, so shouldn't affect anything prior

## Review Notes / Questions

Ready to merge! 🥳 

---

Fixes https://github.com/directus/directus/issues/20559

rijkvanzanten committed:
Rename tag directus-custom to directus-extension

rijkvanzanten committed:
Devil's in the details

rijkvanzanten committed:
[WIP] Start on search lib

rijkvanzanten committed:
Setup extensions-registry package

rijkvanzanten committed:
Update pnpm lock

rijkvanzanten committed:
Move registry search fn to new package

rijkvanzanten committed:
Squash bugs

paescuj committed:
Remove unused deps

paescuj committed:
Remove unused dep in `extensions`

paescuj commented on packages/extensions/src/shared/constants/extension-keywords.ts:
```diff
@@ -0,0 +1 @@
+export const KEYWORD_EXTENSION = 'directus-extension';
```
Are we going to make use of this in `extensions-sdk` & `extensions-registry`?

rijkvanzanten replied:
Apparently this constant isn't used anywhere at all and there's two keywords, so I'll drop this one for now

paescuj committed:
Merge branch 'next' into forum-browse

rijkvanzanten committed:
Add tests for validate-limit

rijkvanzanten committed:
Finish test coverage

rijkvanzanten committed:
Remove unused constant

rijkvanzanten committed:
Install extensions-registry

rijkvanzanten committed:
Add default keywords

rijkvanzanten committed:
Expose registry endpoint in rest api

paescuj committed:
Add back lodash-es 😇

rijkvanzanten committed:
Add describe module

rijkvanzanten committed:
Reinstal pnpm

rijkvanzanten committed:
Drop reliance on author field

The problem is that there's no requirement for it to be a valid(ated) user. Publisher is the only field we can rely on being an existing NPM user.

paescuj commented on packages/extensions-registry/src/modules/search/lib/construct-text.ts:
```diff
@@ -0,0 +1,28 @@
+		 *
+		 * @TODO remove `directus-custom-x` tag search
+		 */
+		text += ` keywords:"directus-extension-${options.extensionType}","directus-custom-${options.extensionType}"`;
```
Apparently, quotes need to be around the whole keywords set:
```suggestion
		text += ` keywords:"directus-extension-${options.extensionType},directus-custom-${options.extensionType}"`;
```

https://registry.npmjs.org/-/v1/search?text=keywords:%22directus-extension-bundle%22,%22directus-custom-bundle%22
vs
https://registry.npmjs.org/-/v1/search?text=keywords:%22directus-extension-bundle,directus-custom-bundle%22

paescuj commented on packages/extensions-registry/src/utils/get-package-extension-type.ts:
Since we cannot control the keywords, it think we should fail/ignore packages which have multiple `directus-extension-` keywords as we cannot correctly categorize them.
Right now, there are such packages out in the wild 😢 

In addition, it might be worth adding a check to `extensions-sdk` (runs on `add`, `build` commands?) which ensures:
- there's only one `directus-extension-` keyword
- the keyword corresponds to the type

paescuj committed:
Update convertSearchResult test after author change

paescuj committed:
Format

paescuj commented on packages/extensions-registry/src/modules/search/utils/validate-text.ts:
```diff
@@ -0,0 +1,14 @@
+/**
+ * Throw a TypeError if the search text contains an npm special search qualifier
+ */
```
```suggestion
/**
 * Ensures the search text doesn't contain any npm special search qualifier
 * @throws TypeError if text contains npm special search qualifiers
 */
```

rijkvanzanten committed:
Allow anything after the route

To support packages with scopes

rijkvanzanten committed:
Add test for describe

rijkvanzanten committed:
Add more tests

rijkvanzanten committed:
Throw unprocessablecontenterror instead of error

rijkvanzanten committed:
Install errors pkg

rijkvanzanten committed:
Finish test coverage

rijkvanzanten committed:
Don't require description

rijkvanzanten committed:
Don't wrap individual keywords

rijkvanzanten committed:
Merge branch 'next' into forum-browse

rijkvanzanten committed:
Run formatter

rijkvanzanten committed:
Add changeset

rijkvanzanten committed:
Update packages/extensions-registry/src/modules/search/utils/validate-text.ts

Co-authored-by: Pascal Jufer <pascal-jufer@bluewin.ch>

rijkvanzanten committed:
Return sandbox information

rijkvanzanten committed:
Drop support for directus-custom keyword

rijkvanzanten committed:
Set type to null for multiple tags

rijkvanzanten committed:
Restart api for new registry

rijkvanzanten committed:
Merge branch 'next' into forum-browse

rijkvanzanten committed:
Add version utility

rijkvanzanten committed:
Add out-of-date error

rijkvanzanten committed:
Add assertion for api versions

rijkvanzanten committed:
Finish new list method

rijkvanzanten committed:
Add describe

rijkvanzanten committed:
Use updated registry functions

rijkvanzanten committed:
Fix controller usage

rijkvanzanten committed:
Return output in data flag to be consistent

rijkvanzanten committed:
Add tests for the describe module

rijkvanzanten committed:
Add tests for list module

rijkvanzanten committed:
Finish tests for extensions-registry package

rijkvanzanten committed:
Remove vscode shenanigans

DanielBiegler commented on packages/extensions-registry/package.json:
```diff
@@ -0,0 +1,46 @@
+{
+	"name": "@directus/extensions-registry",
+	"version": "0.0.0",
```
```suggestion
	"version": "0.1.0",
```

Not sure if its important but wanted to mention just in case

paescuj replied:
Looks strange but `0.0.0` is indeed correct, as it will be bumped by changeset 👍

azrikahar reviewed (approved):
Code LGTM 👍
